flax/linen/normalization.py

```diff
@@ -849,3 +849,169 @@ def _spectral_normalize(self, path, vs, layer_instance_name, update_stats):
+    param_dtype: the dtype passed to parameter initializers (default: float32).
+    use_scale: If True, creates a learnable variable ``scale`` that is
+      multiplied to the ``layer_instance`` variables after l2-normalization.
+    scale_init: Initialization function for the scaling function.
+    feature_axes: The feature axes dimension(s). The l2-norm is calculated by
+      reducing the ``layer_instance`` variables over the remaining (non-feature)
+      axes. Therefore a separate l2-norm value is calculated and a separate
+      scale (if ``use_scale=True``) is learned for each specified feature. By
+      default, the trailing dimension is treated as the feature axis.
+    variable_filter: An optional iterable that contains string items. The
+      WeightNorm layer will selectively apply l2-normalization to the
+      ``layer_instance`` variables whose key path (delimited by '/') has a
+      match with ``variable_filter``. For example, ``variable_filter={'kernel'}``
+      will only apply l2-normalization to variables whose key path contains
+      'kernel'. By default, ``variable_filter={'kernel'}``.
+  """
+
+  layer_instance: Module
+  epsilon: float = 1e-12
+  dtype: Optional[Dtype] = None
+  param_dtype: Dtype = jnp.float32
+  use_scale: bool = True
+  scale_init: Callable[[PRNGKey, Shape, Dtype], Array] = initializers.ones
+  feature_axes: Optional[Axes] = -1
+  variable_filter: Optional[Iterable] = field(default_factory=lambda: {'kernel'})
+
+  @compact
+  def __call__(self, *args, **kwargs):
+    """Compute the l2-norm of the weights in ``self.layer_instance``
+    and normalize the weights using this value before computing the
+    ``__call__`` output.
+
+    Args:
+      *args: positional arguments to be passed into the call method of the
+        underlying layer instance in ``self.layer_instance``.
+      **kwargs: keyword arguments to be passed into the call method of the
+        underlying layer instance in ``self.layer_instance``.
+
+    Returns:
+      Output of the layer using l2-normalized weights.
+    """
+
+    def layer_forward(layer_instance):
+      return layer_instance(*args, **kwargs)
+
+    return map_variables(
+        layer_forward,
+        trans_in_fn=lambda vs: jax.tree_util.tree_map_with_path(
+          self._l2_normalize,
+          vs,
+        ),
+        init=self.is_initializing(),
+    )(self.layer_instance)
+
+  def _l2_normalize(self, path, vs):
+    """Compute the l2-norm and normalize the variables ``vs`` using this
+    value. This is intended to be a helper function used in this Module's
+    ``__call__`` method in conjunction with ``nn.transforms.map_variables``
+    and ``jax.tree_util.tree_map_with_path``.
+
+    Args:
+      path: dict key path, used for naming the ``scale`` variable
+      vs: variables to be l2-normalized
+    """
+    value = jnp.asarray(vs)
+    str_path = self.layer_instance.name + '/' + '/'.join((dict_key.key for dict_key in path[1:])) + '/scale'
```

Comment:
By adding `scale` at the end of the path here, all variables will be normalized if the user is trying to normalize the `scale` variable.

Reply:
Should I keep the keypath and just remove `scale` from the end of it, or should I rename it to something like `weight_scale`?